SalesforceCommerceCloud/pwa-kit
# Spike: Server-side Hooks `useProps`

[see quip doc for more discussion](https://salesforce.quip.com/YlHbAXXHVbQN)

# Description

This PR is a potential solution to the GUS ticket [Support Server-Side React.js Hook Support](https://gus.lightning.force.com/a07EE00000wjAz2YAE).

The primary goal of this PR is to be able to support running hooks on the server. Specifically hooks used to retrieve data from API's, like SAPI and 1C. Although we haven't done any work converting the `commerce-api` folder in the _retail react app_ template into its own hooks library, we are imagining that some of the hooks might look something like `useBasket`, `useProduct`, `useCategory` etc. So in order to use those hooks on the server, to support server-side rendering, we first must extend the API of our SDK to allow use to run these types of hooks. 

The secondary goal of this PR is to create a solution that seems like a natural progression from our current solution `getProps`. I didn't want a completely different API than what we already had, and I want developers to say "hey that was easy" when converting their `getProps` projects to hooks. 

## Solution

This PR attempts to solve the problem of server-side hooks by introducing a special hook provided by the SDK called `useProps`. Although hooks in general aren't run when rending on the server, you can craft things in a way that allows you to track what hooks are instantiated during the `renderToString` call. We do this using a context, which we will then take all the tracked requests from those hooks and invoke them. Below is an example of a simple component that uses this new hook to get data and render is on the server:

```
const MyComponent = () => {
    // This example uses a synchronous function, but works equally as well with asynchronous functions.
    const {person} = useProps(() => ({
        person: {name: 'Ben'}
    }))

    return (
        <div>Hello {person.name}<div>
    )
}
```

Building on this example, lets take a look at what it might looks like to get data from an rest API:

```
const MyProductComponent = () => {
    const api = useCommerceAPI() // This will be added via `extraGetPropsArgs` in a future iteration.

    // This example uses a synchronous function, but works equally as well with asynchronous functions.
    const {product} = useProps(async ({params}) => {
        const product = await api.getProduct(params.productId)

        return {product}
    })

    return (
        <div>
            <div>Name: {product.name}<div>
            <div>Price: {product.price}<div>
        <div>
    )
}
```

You can see that we can successfully use this method to get data from remote APIs, but whats also interesting is the the arguments passed to your `didUpdate` function look very similar to the arguments passed to `getProps`. These include, the request and response objects, the location and params, and the extra args defined in the app configurations `getExtraGetPropsArgs`. 

Here is what the signature of the `useProps` functions looks like.

```
/**
 * This hook runs the `didUpdate` function both on client and server, 
 * any values returned from this function can be used in rendering.
 *
 * @param {Any} initialValue - Inital values returned, these can be used for 
 *  loading/placeholder state data.
 * 
 * @param {Function} didUpdate - The function to be run whenever the source values
 *  change. This function is required to return a value.
 * 
 * @param {Array} source - The state/props to that cause the `didUpdate` function 
 *  to be run.
 * 
 * @return {Any} - The value returned in `didUpdate`
 * 
 * NOTE: The `didUpdate` function will be passed an `arg` object that consists 
 * of the following values.
 * @param {Object} args
 * @param {Request} args.req
 * @param {Response} args.res
 * @param {Location} args.location
 * @param {Object} args.param
 */
export const useParams = (initialValue, didUpdate, source) => returnVal;
```

NOTE: I have not implemented the adding the the extra args, but I don't imagine that being much of a problem. At the same time I am going to work on the return value to make sure the API is tightened up.

### Where does the data go?

`getProps` data is serialized in the `window.__PRELOADED_STATE__` object under the `appProps` or `pageProps` keys. Similarly the `useProps` data is serialized in the same object but under the key `hooksProps`. Each `useProps` hook "used" in the application for a given url is keyed by a unique identifier starting with `uh`. What this means, is that you can use multiple instances of these hooks on a single page. This give you a little more flexibility with how your `didUpdate` function reacts to the `source` values.

Here is what the serialized data look like loading the home page (note there is some testing data in there), but the key to this is that we no longer have `pageData` as we deleted the `getProps` definition. Instead that data is under the `hookProps` key:

![image](https://user-images.githubusercontent.com/8902581/170771492-54fd3c8a-295c-42f4-8095-d358695dd0e7.png)

## Caveats

No solution is perfect. One of the major changes required to get his one working, is that it's required that we render our application twice, once to determine what server-side hooks are used, then one to render with that data. Previously we didn't have to do that because we knew where to find the `getProps` functions (they were dangling off the App and Page classes). Another issue I'm seeing is about setting header values, although I tested setting the status of a response (and it works) the initial prototype has issues setting headers, this is blocking, but at the same time I think this is something that is solvable, and I just need time to fix it.

## Benefits

Once upside to this solution is that you can use the `useProps` hook anywhere in you app, not just on the pages and App components like `getProps`. This means that you don't have to drill down props in to other components that require you to wire up data. Pretty neat.

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# TODOs

- [x] Currently the didUpdate functions are being run twice when rendering on the server due to the double render. Figure out a way to no immediately invoke those functions.

# Changes

- Added new `useProps` hook, that works in conjunction with a `use-props-context`.
- Modified react rendering to render twice and record all use props didUpdate requests.
- Port home and product pages to use new hook.

# How to Test-Drive This PR

- Checkout branch and see that homepage and pdp work (note: pdp needs some tweaks) in that they are getting data on the server.
